### PR TITLE
Permalinks should be able to decode and encode via servers.

### PR DESCRIPTION
**Pull request by Gnuxie**
Previously it was skipping `&via` and just joining `via` when there were more servers. And it also just would not attempt to parse via servers when there was no event id to go with a room.

Closes https://github.com/turt2live/matrix-bot-sdk/issues/299

## Checklist

* [x] Tests written for all new code
* [x] Linter has been satisfied
* [x] Sign-off given on the changes (see CONTRIBUTING.md)

### CODE REVIEW

File: test/helpers/PermalinksTest.ts [hunk 1]
```diff
@@ -4,27 +4,27 @@ describe('Permalinks', () => {
     describe('forRoom', () => {
         it('should generate a URL for a room ID', () => {
             const roomId = "!test:example.org";
-            const expected = `https://matrix.to/#/${roomId}`;
+            const expected = `https://matrix.to/#/${encodeURIComponent(roomId)}`;
```
Comment (turt2live) [inline test/helpers/PermalinksTest.ts:7]: not encoding is deliberate for compatibility with some clients: the SDK should be able to parse encoded and unencoded URLs, but should not be producing encoded URLs.

Comment (Gnuxie) [reply test/helpers/PermalinksTest.ts]: ok, that's no problem, I've changed it

> Historically, clients have not produced URIs which are fully encoded. Clients should try to interpret these cases to the best of their ability. For example, an unencoded room alias should still work within the client if possible.

> The components of the matrix.to URI (<identifier> and <extra parameter>) are to be percent-encoded as per RFC 3986

Though the spec does imply the bad links should no longer be generated when it probably should not.

File: test/helpers/PermalinksTest.ts [hunk 2]
```diff
@@ -125,26 +129,37 @@ describe('Permalinks', () => {
...
             const via = ["one.example.org", "two.example.org"];
             const expected: PermalinkParts = { userId: undefined, roomIdOrAlias: roomId, viaServers: via, eventId };
-            const parsed = Permalinks.parseUrl(`https://matrix.to/#/${roomId}/${eventId}?via=${via.join("via=")}`);
+            const parsed = Permalinks.parseUrl(`https://matrix.to/#/${roomId}/${eventId}?via=${via.join("&via=")}`);
 
             expect(parsed).toMatchObject(<any>expected);
+            expect(Permalinks.parseUrl(Permalinks.forEvent(roomId, eventId, via))).toMatchObject(expected);
+        });
+
+        it('should parse room ID permalink URLs with via servers', () => {
```
Comment (turt2live) [inline test/helpers/PermalinksTest.ts:156]:
```suggestion
        it('should parse room ID URLs with via servers', () => {
```

(permalinks are technically ones with event IDs)

**Review by turt2live — APPROVED** [review]
I guess I need to actually hit the approve button, hey

**Comment by Gnuxie** [comment]
Thanks!

**Pull request merged**